src/Raven.Server/ServerWide/ClusterStateMachine.cs

```diff
@@ -979,29 +1028,29 @@ private void RemoveNodeFromCluster(TransactionOperationContext context, Blittabl
             }
         }
 
-        private void ExecuteManyOnDispose(TransactionOperationContext context, long index, string type, List<Action> actions)
+        private void ExecuteManyOnDispose(TransactionOperationContext context, long index, string type, List<Func<Task>> tasks)
         {
-            context.Transaction.InnerTransaction.LowLevelTransaction.OnDispose += transaction =>
+            context.Transaction.InnerTransaction.LowLevelTransaction.AfterCommitWhenNewReadTransactionsPrevented += () =>
             {
-                if (transaction is LowLevelTransaction llt && llt.Committed)
-                {
-                    _rachisLogIndexNotifications.AddTask(index);
-                    var count = actions.Count;
+                _rachisLogIndexNotifications.AddTask(index);
+                var count = tasks.Count;
 
-                    if (count == 0)
-                    {
-                        NotifyAndSetCompleted(index);
-                        return;
-                    }
+                if (count == 0)
+                {
+                    NotifyAndSetCompleted(index);
+                    return;
+                }
 
+                context.Transaction.InnerTransaction.LowLevelTransaction.OnDispose += tx =>
+                {
                     var exceptionAggregator =
                         new ExceptionAggregator(_parent.Log, $"the raft index {index} is committed, but an error occured during executing the {type} command.");
 
-                    foreach (var action in actions)
+                    foreach (var task in tasks)
                     {
-                        TaskExecutor.Execute(_ =>
+                        Task.Run(async () =>
```

Comment:
This is going to block until the task is done running?

Reply:
No, it should execute the action inside the `Task.Run` asynchronously.